feat: add schemas for new oob firmware

maxd-nordic opened the pull request.

maxd-nordic committed:
feat: add schemas for new oob firmware

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>

maxd-nordic commented:
I can recommend https://cbor.nemo157.com/ to look into the cbor bytes.

coderbyheart reviewed (changes requested):
Please provide unit tests that ensure that the schemas are valid CDDL and the provided proto messages are valid according to the schemas

coderbyheart commented on .github/workflows/build-and-publish.yaml:
```diff
@@ -12,8 +12,40 @@ env:
   GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 
 jobs:
+  validate:
```
```suggestion
  validate-cddl:
```

coderbyheart commented on .github/workflows/build-and-publish.yaml:
```diff
@@ -12,8 +12,40 @@ env:
+
+      - name: Validate CDDL and convert JSON
+        working-directory: proto/hello/firmware
+        run: |
```
That should be a script file so it can be executed without using GitHub Actions.

maxd-nordic replied:
it now lives in hello/firmware/validate.sh

coderbyheart commented:
Looks great! I think we should maintain these in a separate repo until we have firmware that sends it and the nRF Cloud side ready for receiving and forwarding it via the message bridge. Because right now there is no way to use these in a real setup so it's a protocol that's can't be supported right now, but the `proto` repo is the authors tive source for the protocol in production.

I'll do the move.

coderbyheart commented:
After pondering a bit about it, I think it's better if we keep the future protocols in the same repo. This will allow us to sync changes happening in the MQTT protocol directly in one PR.

coderbyheart committed:
docs: add README

coderbyheart committed:
refactor: use cloudToDevice / deviceToCloud

This is the naming convention used in nRF Cloud API docs.

Use system temp dir for test files

coderbyheart committed:
ci: print JSON representation

coderbyheart commented on hello/firmware/cloudToDevice/led.json:
```diff
@@ -0,0 +1 @@
+[[1, 1688484111, true]]
```
**question**: @maxd-nordic why are the JSON messages always wrapped in an array?

maxd-nordic replied:
All the messages are "from-device-message" or "to-device-message". They are always an array, even if only containing one element.

maxd-nordic replied:
this is to reduce encoding and decoding complexity, and the usual case will be multiple elements

coderbyheart replied:
The "usual case"? Can you provide an example? Why would the device send multiple temperature readings, button presses, or battery percentages? At any given time there is just one of these.

maxd-nordic replied:
The usual case as I picture it is the following: Due to an event or timeout, the device has sampled data from its sensors and sends it to cloud. Since it's more than one sensor, this will be a list of data points.
Pressing the button is a special case where probably only one element is present (although we might even have more data to send at that moment).

coderbyheart replied:
Ah, I see. So this is a deviation from asset_tracker_v2's behaviour where it sends individual messages via MQTT.
